torch/csrc/distributed/c10d/socket.cpp

```diff
@@ -0,0 +1,818 @@
+            port_);
+
+    if (tryConnect(AF_INET)) {
+      return std::move(socket_);
+    }
+  } else {
+    logInfo("The client socket will attempt to connect to an IPv4 or IPv6 address of ({}, {}).",
+            host_,
+            port_);
+
+    if (tryConnect(AF_UNSPEC)) {
+      return std::move(socket_);
+    }
+  }
+
+  std::string msg{};
+  if (timed_out_) {
+    msg = "The client socket has timed out while waiting to connect to ({}, {}).";
+  } else {
+    msg = "The client socket has failed to connect to any network address of ({}, {}).";
+  }
+
+  msg = fmt::format(msg, host_, port_);
+
+  logError(msg);
+
+  msg = fmt::format("{} {}", msg, fmt::join(errors_, " "));
+
+  if (timed_out_) {
+    throw TimeoutException{msg};
+  } else {
+    throw SocketError{msg};
+  }
+}
+
+bool SocketConnectOp::tryConnect(int family) {
+  ::addrinfo hints{}, *naked_result = nullptr;
+
+  hints.ai_flags = AI_ADDRCONFIG | AI_V4MAPPED | AI_ALL | AI_NUMERICSERV;
+  hints.ai_family = family;
+  hints.ai_socktype = SOCK_STREAM;
+
+  int r = ::getaddrinfo(host_, port_.c_str(), &hints, &naked_result);
+  if (r != 0) {
+    const char* gai_err = ::gai_strerror(r);
+
+    recordError("The {}network addresses of ({}, {}) cannot be retrieved (gai error: {} - {}).",
+                family == AF_INET ? "IPv4 " : family == AF_INET6 ? "IPv6 " : "",
+                host_,
+                port_,
+                r,
+                gai_err);
+
+    return false;
+  }
+
+  addrinfo_ptr result{naked_result};
+
+  for (::addrinfo* addr = naked_result; addr != nullptr; addr = addr->ai_next) {
+    logInfo("The client socket is attempting to connect to {}.", *addr);
+    if (tryConnect(*addr)) {
+      return true;
+    }
+  }
+
+  return false;
+}
+
+bool SocketConnectOp::tryConnect(const ::addrinfo& addr) {
+  TimePoint deadline = Clock::now() + opts_->connect_timeout();
+
+  while (Clock::now() < deadline) {
+    SocketImpl::Handle hnd = ::socket(addr.ai_family, addr.ai_socktype, addr.ai_protocol);
+    if (hnd == SocketImpl::invalid_socket) {
+      recordError("The client socket cannot be initialized to connect to {} {}.",
+                  addr,
+                  getSocketError());
+
+      return false;
+    }
+
+    socket_ = std::make_unique<SocketImpl>(hnd);
+
+    socket_->enableNonBlocking();
+
+    // This function returns -1 in case of an error, 0 if the attempt has timed
+    // out, and 1 if the connection is established.
+    int r = tryConnect(addr, deadline);
+    if (r == -1) {
+      std::error_code err = getSocketError();
+      if (err == std::errc::interrupted) {
+        throw std::system_error{err};
+      }
+
+      // Retry if the server is not yet listening or if its backlog is exhausted.
+      if (err == std::errc::connection_refused ||
+          err == std::errc::connection_reset ||
+          err == std::errc::connection_aborted) {
+        logWarning("The server socket on {} is not yet listening {}, retrying...", addr, err);
```

Comment:
it looks like we keep retrying until timeout, is it worth it to also log how many times we've now retried?

Reply:
From user's perspective we try it only once. These three error conditions only happen if the server is not reachable. Once it becomes reachable (we internally retry every second), the `connect()` syscall will take the rest of the function and use `poll()` with the remaining timeout value. So there is really not a clear way to communicate what we mean by an "internal retry". If the server is reachable, but `connect()` hangs, we may not even have a retry.